### PR TITLE
Fix :include: filter to require all patterns to match. Fix stacktrace…

### PR DESCRIPTION
**Pull request by amykyta3**
## Summary of the changes / Why this is an improvement
* Fix logic of `:include:` filter to match original intent. Include filter should only include the row if _all_ filters match. Updated readme to describe behavior.
* Fix stacktrace if a column filter fetches an empty cell. Sanitizing to an empty string so that the user still has the option of matching against it with a regex.

## Checklist

 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed

### CODE REVIEW

**Comment by chaudum** [comment]
Hi @amykyta3 , we'll have a look!  Would you mind writing a test that covers the described issue?

**Comment by chaudum** [comment]
retest this please

File: src/crate/sphinx/csv.py [hunk 1]
```diff
@@ -70,27 +70,27 @@ def _apply_filters(self, rows, max_cols, include_filters, exclude_filters):
...
-                            break
+            for col_idx, pattern in include_filters.items():
+                # cell data value is located at hardcoded index pos. 3
+                # data type is always a string literal
+                if max_cols - 1 >= col_idx:
+                    # sanitize to empty string if cell is empty
+                    cell = row[col_idx][3][0] if row[col_idx][3] else ""
+                    if not pattern.match(cell):
+                        # ... unless any of the include filters do not match
+                        # it's cell ...
```
Comment (MarkusH) [inline src/crate/sphinx/csv.py]:
```suggestion
                        # its cell ...
```

File: src/crate/sphinx/csv.py [hunk 2]
```diff
@@ -70,27 +70,27 @@ def _apply_filters(self, rows, max_cols, include_filters, exclude_filters):
...
+                        include = False
+                        break
 
             # ... unless exclude filters are defined (as well) ...
```
Comment (MarkusH) [inline src/crate/sphinx/csv.py]:
```suggestion
             # ... or unless exclude filters are defined (as well) ...
```

File: README.rst [hunk 3]
```diff
@@ -47,11 +47,13 @@ This plugin adds the following options to the csv-table_ directive:
     This option takes a Python dict specifying the column index (starting at
     zero) and a regular expression. Rows are included if the columnar value
     matches the supplied regular expression.
+    If multiple filters are given, row is included if all columns match.
 
 ``:exclude:``
     This option takes a Python dict specifying the column index (starting at
     zero) and a regular expression. Rows are excluded if the columnar value
     matches the supplied regular expression.
+    If multiple filters are given, row is excluded if any columns match.
```
Comment (MarkusH) [inline README.rst]: This is not the case, if I'm not mistaken. We exclude when _any_ exclude filter matches, not when _all_ exclude filter match. And I'd like to keep it that way.

To make the difference to `:include:` clear, let's rephrase the last sentence:

`Rows are excluded if the columnar value matches *any* supplied regular expression.`

File: README.rst [hunk 4]
```diff
@@ -47,11 +47,13 @@ This plugin adds the following options to the csv-table_ directive:
     This option takes a Python dict specifying the column index (starting at
     zero) and a regular expression. Rows are included if the columnar value
```
Comment (MarkusH) [inline README.rst]: Let's rephrase the sentence as is:

`Rows are included if the columnar value matches *all* supplied regular expression.`

**Comment by MarkusH** [comment]
ok to test

**Comment by amykyta3** [comment]
Added to testcase to cover empty cell issue. Also added tests that verify multiple include/exclude filters.

**Comment by amykyta3** [comment]
A testcase has been added. Is there anything else I need to do from my end?

**Comment by MarkusH** [comment]
HI @amykyta3, sorry for not replying earlier. Given the change in behavior, we involved our product team. They're evaluating the way forward.

**Comment by seut** [comment]
ok to test

**Comment by seut** [comment]
@amykyta3 
After some discussion we've come to the conclusion that we would not like to get this merge cause of the behaviour change of `include -> ANY` vs `include -> ALL`.
Both use-cases are valid and right now, we prefer to be consistent with the `exclude` filter behaviour.

I think we should rather go into letting the user define this, maybe using https://pandas.pydata.org/ together with a possibility to configure queries/filter e.g. using SQL (https://github.com/yhat/pandasql/).

We really appreciate your effort, even if not merged, it definitely helps gathering ideas how this could be achieved.
Would you mind creating a dedicated PR for the empty cell fix? We'd love to get this in.
Thank you a lot.

**Pull request closed without merging**